README.md
# FFMPEG PHP extension

This project was originally released on sourceforge.net. This distribution
includes patches applied by Townnews.com throughout the years to keep the
extension operational.

It previously only worked with FFMPEG 1.2.12 or older but has been modified
so that it compiles now against FFMPEG 3.4.0 (Cantor). This extension has
been tested against:

* PHP 5.6
* PHP 7.2.6

## Configuration

Some custom expansions on this version include support for setting some
resource limits via INI:

```ini
; Load the PHP extension
extension=ffmpeg.so

; Maximum threads used for processing videos
ffmpeg.threads=1

; Memory allocation size in bytes
ffmpeg.max_alloc=67108864
```


ChangeLog
ChangeLog for ffmpeg-php (ffmpeg-php.sourceforge.net)

version 0.9.0 (6/19/2018)
   Changes:
   * Updated with PHP 7.2 support (contributed by dward)
   * Fixed unit tests. Made getAuthor() check 'artist' and getYear()
     check 'date' metadata fields

version 0.8.0 (11/16/2017)
   Changes:
   * Modified to compile and work with FFMPEG 3.4.0 (#1)
   * Updated README documentation

version 0.7.0 (09/22/2014)
	New Features:
	* Support for new INI directive: ffmpeg.max_alloc=<bytes> to set a
	  memory cap on the usage of FFMPEG
	* Support for new INI directive: ffmpeg.threads=<threads> to set the
	  maximum number of threads that can be used by FFMPEG
	* Made compatible with PHP 5.4
	* Made compatible with FFMPEG 1.2.4

	Bug Fixes:
	* Fixed issue with config.m4 and alternate libdir

version 0.6.0 (10/15/2008)
    New Features:
    * Support for ffmpeg swscale API

    API Changes:
    * Removed obsolete resize() and crop() functions. These operations are
      more flexibly done using the gd commands.
    * Removed animated gif class since it was way out of date and the 
      quality of animated gifs created by ffmpeg sucks anyway

    Bug Fixes:
    * Added check to ensure that gd loads before ffmpeg-php if both are
      compiled as extensions.
    * Replaced large mp3 test file with smaller sample to avoid copyright
      issues and reduce package size.
    * Convert from deprecated img_* functions to swscale.
    * Builds against php4 correctly again.

version 0.5.3 (05/30/2008)
    New Features:
    * Added 'ffmpeg.show_warnings' ini flag to control whether ffmpeg 
      warnings are converted to php warnings.
    * Applied patch to add getAudioStreamId and getVideoStreamId functions 

version 0.5.2.1 (04/03/2008)

    Bug Fixes:
    * Minor build fixes and cleanups

version 0.5.2 (04/02/2008)

    Bug Fixes:
    * ffmpeg-php should compile on macos-x now
    * Failure to find the proper codec is not a fatal error anymore
    * Updated build script to deal with new ffmpeg include file layout.

    
    New Features:
    * Added gd image bounds checking.
    * Added config flag to skip the compile-time gd check.
    * ffmpeg's logs are mapped to php's warnings and notices (no more stderr pollution)

version 0.5.1 (06/11/2007)

    Bug Fixes:
    * Fixed some of the tests (Thanks to Raphael Geissert for the patch)
    * Improved build process (Thanks to Raphael Geissert for the suggestions)
    * Fixed the test script so it won't generate warnings when video/audio
      stream is missing.

    New Features:
    * Added $movie->hasVideo() function to test for video stream.
    * Copyright updated for better compatibilty with the PHP license.

version 0.5.0 (09/29/2006)

    Bug Fixes:
    * Fixed problem with params to av_open_input_file not getting initialized
      correctly. (Thanks to SilverCode for the fix)

version 0.4.9 (09/04/2006)

    New Features:
    * Better error reporting when opening a movie fails
    * INSTALL doc clarifications

    Bug Fixes:
    * Fixed compilation error in ffmpeg_animated_gif.c when building against 
      old ffmpeg


version 0.4.8 (08/04/2006)
    
    New Features:
    * Added new functions for audio streams: getAudioSampleRate(), 
      getAudioBitRate() (Martin Prangl)
    * Changed getBitRate function to like getAudioSampleRate() and
      getAudioBitRate(). Now returns bits per second instead of
      calculating kb/s. (To get kb/s just do movie->getBitRate() / 1000
      in php.) (tkirby)

    Bug Fixes:
    * Fix possible error in hasAudio calculation if audio stream is the
      first stream. (tkirby)

version 0.4.7 (06/08/2006)

    Bug Fixes:
    * Fixed framerate calculation to be compatible with ffmpeg cvs 6/8/06 
      (tkirby)

version 0.4.6 (08/30/2005)

    Bug Fixes:
    * Fixed compatiblity with ffmpeg cvs after 7/30/05 or so (tkirby)

version 0.4.5 (07/12/2005)
    Bug Fixes:
    * Minor build fixes for older ffmpeg versions (tkirby)

version 0.4.4 (06/24/2005)
    New Features:
    * Added ability to create animated gifs via new ffmpeg_animated_gif
      class (tkirby)
    * Fixed ffmpeg_frame constructor to allow creating ffmpeg_frames from
      gd images. This allows to make animated gifs directly from a series
      of gd images. I believe this is currently the only way to make 
      animated gifs using PHP since the bundled version of gd does not
      support animated gif creation (tkirby)
    * Added flag to php.ini to allow server admins to disable persistent
      movies since injudicious use of persistent movies in scripts can eat 
      up server resources quickly (tkirby)
    * Added frame->isKeyFrame() and movie->getNextKeyFrame() methods (tkirby)
    * Now displays ffmpeg-php ini values in php_info() table (tkirby)

    Bug Fixes:
    * Fixed parsing of 'persistent' boolean flag in ffmpeg_movie constructor
    * Fixed segfault when a script tries to read more frames than exist in
      a movie. (tkirby)
    * Patch to use native time bases. This allows building against libavcodec 
      build 4754+ (Steve Brown)
    * Patch to corrctly init gd before attempting to fetch gd image resources.
      This fixes the "togdimage() supplied resource is not a valid Image 
      resource" error that a few people have reported (Steve Brown)

version 0.4.3 (03/12/2005)
    Bug Fixes:
    * Added ability to open movies as persistent resources. This greatly
      increases speed of repeatedly accessing the same movie from separate
      php scripts. This is still somewhat experimental so don't use it
      unless you have a specific need (tkirby)
    * Increased strictness of compiler warnings to catch issues like
      undefined symbol errors. (tkirby)
    * Track frame number within ffmpeg-php since the frame_number variable
      in ffmpeg's decoder context seems to give inaccurate results on some 
      3GP movies. (tkirby)
    * Fixed bug #1099229 - Stride problems if ffmpeg_frame is resized upwards
      (tkirby)

php_ffmpeg.h
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

 */

#ifndef PHP_FFMPEG_H
#define PHP_FFMPEG_H

#include "php_version.h"

/* 
 * PHP-4 doesn't have the METHOD, ME and MALIAS macros so map them back
 * to the function macros for PHP-4
 */
#if PHP_MAJOR_VERSION <= 4
#define FFMPEG_PHP_METHOD(a,b) PHP_FUNCTION(b)
#define FFMPEG_PHP_ME(a,b,c,d) PHP_FE(a,c)
#define FFMPEG_PHP_MALIAS(a,b,c,d,e) PHP_FALIAS(b,c)
#else /* Use METHOD macros for PHP-5 */
#define FFMPEG_PHP_METHOD(a,b) PHP_METHOD(a,b)
#define FFMPEG_PHP_ME(a,b,c,d) PHP_ME(a,b,c,d)
#define FFMPEG_PHP_MALIAS(a,b,c,d,e) PHP_MALIAS(a,b,c,d,e)
#endif


#if PHP_MAJOR_VERSION < 7
#define FFMPEG_ZVAL_STRING(zv, str, persist) ZVAL_STRING(zv, str, persist)
#define FFMPEG_REGISTER_RESOURCE(result, ptr, type) ZEND_REGISTER_RESOURCE(result, ptr, type)
#define FFMPEG_RESOURCE zend_rsrc_list_entry
#define FFMPEG_RETURN_STRINGL(str, len, persist) RETURN_STRINGL(str, len, persist)
#else
#define FFMPEG_ZVAL_STRING(zv, str, persist) ZVAL_STRING(zv, str)
#define FFMPEG_REGISTER_RESOURCE(result, ptr, type) zend_register_resource(ptr, type)
#define FFMPEG_RESOURCE zend_resource
#define FFMPEG_RETURN_STRINGL(str, len, persist) RETURN_STRINGL(str, len)
#endif

#define SAFE_STRING(s) ((s)?(s):"")

#ifndef safe_emalloc
    # define safe_emalloc(a,b,c) emalloc((a)*(b)+(c))
#endif

extern zend_module_entry ffmpeg_module_entry;
#define phpext_ffmpeg_ptr &ffmpeg_module_entry

#ifdef PHP_WIN32
#define PHP_FFMPEG_API __declspec(dllexport)
#else
#define PHP_FFMPEG_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

PHP_MINIT_FUNCTION(ffmpeg);
PHP_MSHUTDOWN_FUNCTION(ffmpeg);
PHP_RINIT_FUNCTION(ffmpeg);
PHP_RSHUTDOWN_FUNCTION(ffmpeg);
PHP_MINFO_FUNCTION(ffmpeg);


#ifdef ZTS
#define FFMPEG_G(v) TSRMG(ffmpeg_globals_id, zend_ffmpeg_globals *, v)
#else
#define FFMPEG_G(v) (ffmpeg_globals.v)
#endif

#endif	/* PHP_FFMPEG_H */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * indent-tabs-mode: t
 * End:
 */


ffmpeg_frame.h
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

 */

#ifndef FFMPEG_FRAME_H
#define FFMPEG_FRAME_H

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php_version.h"

#include <libavcodec/avcodec.h>
#include <libavformat/avformat.h>
#include <libavutil/avutil.h>

/* object can't be created from user space so no PHP constructor */
//FFMPEG_PHP_METHOD(ffmpeg_frame, __construct);

/* frame methods */
FFMPEG_PHP_METHOD(ffmpeg_frame, getWidth);
FFMPEG_PHP_METHOD(ffmpeg_frame, getHeight);
FFMPEG_PHP_METHOD(ffmpeg_frame, isKeyFrame);
FFMPEG_PHP_METHOD(ffmpeg_frame, getPresentationTimestamp);

#if HAVE_LIBGD20
FFMPEG_PHP_METHOD(ffmpeg_frame, toGDImage);
#endif // HAVE_LIBGD20

typedef struct {
    AVFrame *av_frame;
    int width;
    int height;
    int pixel_format;
    int keyframe;
    int64_t pts;
} ff_frame_context;

ff_frame_context* _php_create_ffmpeg_frame(INTERNAL_FUNCTION_PARAMETERS);

int _php_convert_frame(ff_frame_context *ff_frame, int new_fmt);

#if PHP_MAJOR_VERSION < 7

#define GET_FRAME_RESOURCE(ffmpeg_frame_object, ffmpeg_frame) {\
	zval **_tmp_zval;\
    if (zend_hash_find(Z_OBJPROP_P(ffmpeg_frame_object), "ffmpeg_frame",\
                sizeof("ffmpeg_frame"), (void **)&_tmp_zval) == FAILURE) {\
        zend_error(E_ERROR, "Unable to locate ffmpeg_frame resource in this object.");\
        RETURN_FALSE;\
    }\
\
    ZEND_FETCH_RESOURCE(ffmpeg_frame, ff_frame_context*, _tmp_zval, -1,\
            "ffmpeg_frame", le_ffmpeg_frame);\
}\

#else

#define GET_FRAME_RESOURCE(ffmpeg_frame_object, ffmpeg_frame) {\
    zend_resource *le;\
    if ((le = zend_hash_str_find_ptr(Z_OBJPROP_P(ffmpeg_frame_object), "ffmpeg_frame",\
                sizeof("ffmpeg_frame") - 1)) == NULL) {\
        zend_error(E_WARNING, "Unable to locate ffmpeg_frame resource in this object.");\
        RETURN_FALSE;\
    }\
\
    ffmpeg_frame = le->ptr; \
}\


#endif

#endif // FFMPEG_FRAME_H

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4
 * vim<600: noet sw=4 ts=4
 */


ffmpeg_tools.c
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <php.h>

#include "ffmpeg_tools.h"

#ifdef HAVE_SWSCALER
#include <libswscale/swscale.h>
#endif

/* {{{ ffmpeg_img_convert() 
 * wrapper around ffmpeg image conversion routines
 */
int ffmpeg_img_convert(AVPicture *dst, int dst_pix_fmt,
        AVPicture *src, int src_pix_fmt,
        int src_width, int src_height)
{
#ifndef HAVE_SWSCALER // No SWSCALER so just use img_convert
    return img_convert(dst, dst_pix_fmt, 
            src, src_pix_fmt, src_width, src_height);
#else // Do swscale convert
    int result = 0;
    struct SwsContext *sws_ctx = NULL;

    if (src_pix_fmt == dst_pix_fmt) {
        return 0;
    }

    // TODO: Try to get cached sws_context first
    sws_ctx = sws_getContext(
            src_width, src_height, src_pix_fmt, 
            src_width, src_height, dst_pix_fmt, 
            SWS_BICUBIC, NULL, NULL, NULL);

    if (sws_ctx == NULL){
        return 1;
    }

    result = sws_scale(sws_ctx, 
            (const uint8_t* const*)src->data, src->linesize,
            0, src_height,
            dst->data, dst->linesize);

    sws_freeContext(sws_ctx);

    if (result == 0){
        return 2;
    }
#endif // NOT HAVE_SWSCALER
    return 0;
}
/* }}} */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4
 * vim<600: noet sw=4 ts=4
 */


ffmpeg_tools.h
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php
   give you permission to combine ffmpeg-php with code included in the
   standard release of PHP under the PHP license (or modified versions of
   such code, with unchanged license). You may copy and distribute such a
   system following the terms of the GNU GPL for ffmpeg-php and the licenses
   of the other code concerned, provided that you include the source code of
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the
   code used other than standard release of PHP. If you modify this file, you
   may extend this exception to your version of the file, but you are not
   obligated to do so. If you do not wish to do so, delete this exception
   statement from your version.

 */

#ifndef FFMPEG_TOOLS_H
#define FFMPEG_TOOLS_H

#include <libavcodec/avcodec.h>

int ffmpeg_img_convert(
        AVPicture *dst, int dst_pix_fmt,
        AVPicture *src, int src_pix_fmt,
        int src_width, int src_height);

#endif // FFMPEG_TOOLS_H

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4
 * vim<600: noet sw=4 ts=4
 */


ffmpeg_errorhandler.c
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Alexey Zakhlestin <indeyets@gmail.com>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php 
   give you permission to combine ffmpeg-php with code included in the 
   standard release of PHP under the PHP license (or modified versions of 
   such code, with unchanged license). You may copy and distribute such a 
   system following the terms of the GNU GPL for ffmpeg-php and the licenses 
   of the other code concerned, provided that you include the source code of 
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the 
   code used other than standard release of PHP. If you modify this file, you 
   may extend this exception to your version of the file, but you are not 
   obligated to do so. If you do not wish to do so, delete this exception 
   statement from your version.

 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"
#include <libavcodec/avcodec.h>

/* {{{ ffmpeg_errorhandler()
 */
void ffmpeg_errorhandler(void *ptr, int level, const char *msg, va_list args)
{
	int php_level;
	TSRMLS_FETCH();

	switch (level) {
		case AV_LOG_ERROR:
			php_level = E_WARNING;
		break;

		case AV_LOG_INFO:
		case AV_LOG_DEBUG:
		default:
			php_level = E_NOTICE;
		break;
	}

	php_verror("", "", php_level, msg, args TSRMLS_CC);
}
/* }}} */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4
 * vim<600: noet sw=4 ts=4
 */


ffmpeg-php.c
/*
   This file is part of ffmpeg-php

   Copyright (C) 2004-2008 Todd Kirby (ffmpeg.php AT gmail.com)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

   In addition, as a special exception, the copyright holders of ffmpeg-php 
   give you permission to combine ffmpeg-php with code included in the 
   standard release of PHP under the PHP license (or modified versions of 
   such code, with unchanged license). You may copy and distribute such a 
   system following the terms of the GNU GPL for ffmpeg-php and the licenses 
   of the other code concerned, provided that you include the source code of 
   that other code when and as the GNU GPL requires distribution of source code.

   You must obey the GNU General Public License in all respects for all of the 
   code used other than standard release of PHP. If you modify this file, you 
   may extend this exception to your version of the file, but you are not 
   obligated to do so. If you do not wish to do so, delete this exception 
   statement from your version.

 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"

#include <libavcodec/avcodec.h>
#include <libavformat/avformat.h>

#if HAVE_SWSCALER
#include <libswscale/swscale.h>
#endif

#include "php_ini.h"
#include "php_globals.h"
#include "ext/standard/info.h"

#include "php_ffmpeg.h"

#define FFMPEG_PHP_VERSION "0.9.0"

zend_module_entry ffmpeg_module_entry = {
#if ZEND_MODULE_API_NO >= 20010901
    STANDARD_MODULE_HEADER,
#endif
    "ffmpeg",
    NULL,
    PHP_MINIT(ffmpeg),
    PHP_MSHUTDOWN(ffmpeg),
    NULL,
    NULL,
    PHP_MINFO(ffmpeg),
#if ZEND_MODULE_API_NO >= 20010901
    FFMPEG_PHP_VERSION, 
#endif
    STANDARD_MODULE_PROPERTIES
};


#ifdef COMPILE_DL_FFMPEG
ZEND_GET_MODULE(ffmpeg);
#endif

extern void register_ffmpeg_movie_class(int);
extern void register_ffmpeg_frame_class(int);
extern void ffmpeg_errorhandler(void *ptr, int level, const char *msg, va_list args);

PHP_INI_BEGIN()
    PHP_INI_ENTRY("ffmpeg.allow_persistent", "0", PHP_INI_ALL, NULL)
    PHP_INI_ENTRY("ffmpeg.show_warnings", "0", PHP_INI_ALL, NULL)
    PHP_INI_ENTRY("ffmpeg.max_alloc", "0", PHP_INI_SYSTEM, NULL)
	PHP_INI_ENTRY("ffmpeg.threads", "1", PHP_INI_SYSTEM, NULL)
PHP_INI_END()


/* {{{ php module init function
 */
PHP_MINIT_FUNCTION(ffmpeg)
{
    /* register all codecs */
    av_register_all();
    
    REGISTER_INI_ENTRIES();
 
    if (INI_BOOL("ffmpeg.show_warnings")) {
        av_log_set_callback(ffmpeg_errorhandler);
    } 

	if (INI_BOOL("ffmpeg.max_alloc")) {
		av_max_alloc(INI_INT("ffmpeg.max_alloc"));
	}
   
    register_ffmpeg_movie_class(module_number);
    register_ffmpeg_frame_class(module_number);

    REGISTER_STRING_CONSTANT("FFMPEG_PHP_VERSION_STRING", 
		    FFMPEG_PHP_VERSION, CONST_CS | CONST_PERSISTENT);
    REGISTER_STRING_CONSTANT("FFMPEG_PHP_BUILD_DATE_STRING", 
		    __DATE__ " " __TIME__, CONST_CS | CONST_PERSISTENT);
    REGISTER_LONG_CONSTANT("LIBAVCODEC_VERSION_NUMBER", 
		    avcodec_version(), CONST_CS | CONST_PERSISTENT);
#ifdef LIBAVCODEC_BUILD
    REGISTER_LONG_CONSTANT("LIBAVCODEC_BUILD_NUMBER", 
		    LIBAVCODEC_BUILD, CONST_CS | CONST_PERSISTENT);
#else
    REGISTER_LONG_CONSTANT("LIBAVCODEC_BUILD_NUMBER",
            avcodec_build(), CONST_CS | CONST_PERSISTENT);
#endif

    return SUCCESS;
}
/* }}} */


/* {{{ php module shutdown function
 */
PHP_MSHUTDOWN_FUNCTION(ffmpeg)
{
    // TODO: Free any remaining persistent movies here?
    
    UNREGISTER_INI_ENTRIES();

    return SUCCESS;
}
/* }}} */


/* {{{ php info function
   Add an entry for ffmpeg support in phpinfo() */
PHP_MINFO_FUNCTION(ffmpeg)
{
    php_info_print_table_start();
//    php_info_print_table_header(2, "ffmpeg-php", "enabled");
    php_info_print_table_row(2, "ffmpeg-php version", FFMPEG_PHP_VERSION);
    php_info_print_table_row(2, "ffmpeg-php built on", __DATE__ " "  __TIME__);
#if HAVE_LIBGD20
    php_info_print_table_row(2, "ffmpeg-php gd support ", "enabled");
#else
    php_info_print_table_row(2, "ffmpeg-php gd support ", "disabled");
#endif // HAVE_LIBGD20
    php_info_print_table_row(2, "ffmpeg libavcodec version", LIBAVCODEC_IDENT);
    php_info_print_table_row(2, "ffmpeg libavformat version", LIBAVFORMAT_IDENT);
#if HAVE_SWSCALER
    php_info_print_table_row(2, "ffmpeg swscaler version", LIBSWSCALE_IDENT);
#else 
    php_info_print_table_row(2, "ffmpeg swscaler", "disabled");
#endif
    php_info_print_table_end();

    DISPLAY_INI_ENTRIES();
}
/* }}} */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4
 * vim<600: noet sw=4 ts=4
 */
